state/indexer/sink/psql/schema.sql

```diff
@@ -1,32 +1,83 @@
+  this schema before using the database to index events.
+ */
+
+-- The blocks table records metadata about each block.
+-- The block record does not include its events or transactions (see tx_results).
+CREATE TABLE blocks (
+  rowid      SERIAL PRIMARY KEY,
+
+  height     INTEGER NOT NULL,
+  chain_id   VARCHAR NOT NULL,
+  created_at TIMESTAMPTZ NOT NULL,
+
+  UNIQUE (height, chain_id)
 );
+
+-- Index blocks by height and chain, since we need to resolve block IDs when
+-- indexing transaction records and transaction events.
+CREATE INDEX idx_blocks_height_chain ON blocks(height, chain_id);
+
+-- The tx_results table records metadata about transaction results.  Note that
+-- the events from a transaction are stored separately.
 CREATE TABLE tx_results (
-    id SERIAL PRIMARY KEY,
-    tx_result BYTEA NOT NULL,
-    created_at TIMESTAMPTZ NOT NULL,
-    UNIQUE (tx_result)
+  rowid SERIAL PRIMARY KEY,
+
+  -- The block to which this transaction belongs.
+  block_id INTEGER NOT NULL REFERENCES blocks(rowid),
+  -- The sequential index of the transaction within the block.
+  index INTEGER NOT NULL,
+  -- When this result record was logged into the sink.
+  created_at TIMESTAMPTZ NOT NULL,
+  -- The hex-encoded hash of the transaction.
+  tx_hash VARCHAR NOT NULL,
+  -- The protobuf wire encoding of the TxResult message.
+  tx_result BYTEA NOT NULL,
+
+  UNIQUE (block_id, index)
 );
-CREATE TABLE tx_events (
-    id SERIAL PRIMARY KEY,
-    key VARCHAR NOT NULL,
-    value VARCHAR NOT NULL,
-    height INTEGER NOT NULL,
-    hash VARCHAR NOT NULL,
-    tx_result_id SERIAL,
-    created_at TIMESTAMPTZ NOT NULL,
-    chain_id VARCHAR NOT NULL,
-    UNIQUE (hash, key),
-    FOREIGN KEY (tx_result_id) REFERENCES tx_results(id) ON DELETE CASCADE
+
+-- The events table records events. All events (both block and transaction) are
+-- associated with a block ID; transaction events also have a transaction ID.
+CREATE TABLE events (
+  rowid SERIAL PRIMARY KEY,
+
+  -- The block and transaction this event belongs to.
+  -- If tx_id is NULL, this is a block event.
+  block_id INTEGER NOT NULL REFERENCES blocks(rowid),
+  tx_id    INTEGER NULL REFERENCES tx_results(rowid),
+
+  -- The application-defined type label for the event.
+  type VARCHAR NOT NULL
 );
-CREATE INDEX idx_block_events_key_value ON block_events(key, value);
-CREATE INDEX idx_tx_events_key_value ON tx_events(key, value);
-CREATE INDEX idx_tx_events_hash ON tx_events(hash);
+
+-- The attributes table records event attributes.
+CREATE TABLE attributes (
+   event_id      INTEGER NOT NULL REFERENCES events(rowid),
+   key           VARCHAR NOT NULL, -- bare key
+   composite_key VARCHAR NOT NULL, -- composed type.key
+   value         VARCHAR NULL,
+
+   UNIQUE (event_id, key)
+);
+
+-- A joined view of events and their attributes. Events that do not have any
+-- attributes are represented as a single row with empty key and value fields.
+CREATE VIEW event_attributes AS
```

Comment:
are these materialized? (pardon my pgsql ignorance?)

Reply:
Not as far as I know. PostgreSQL supports the `CREATE MATERIALIZED VIEW` extension, which suggests the default implementation isn't, but I didn't look that closely.

Reply:
Yes, it's just a stored query. I thought about this issue, and I think it should be OK for a couple reasons:

1. The queries we're storing are straightforward, and could easily be optimized with indexes if necessary.
2. If someone does need to investigate perf, they can use the normal PostgreSQL facilities like the query planner to get insight. 

I figured it wasn't as important to make the "intuitive performance" of the queries obvious, but even then the views are not doing anything complicated, so I think it shouldn't hurt. It seemed to me that the important cases to optimize are writes (since that's the majority of queries online) and table-state lookups (since those are part of writing). 

That said, I am open to arguments on the other side 🙂

Reply:
yeah, I think these are fine, and my only arguments against views in favor of just calling the queries directly are: 

- ease of changing it in the future (code changes are easier than schema changes, where we might leave cruft around in the db and have a larger test burden for migrations.) 
- a chance that someone assumes that a view is a table and then then writes a query that joins or aggregates over a view+query in a way that doesn't quite make sense. 

While I suppose there are performance implications to the second, my concern is really just for "expected behavior" and making it easier to maintain/use going forward.

Reply:
That's a reasonable concern, re: schema updates. I would like to make an argument that—even given that legitimate concern—views are probably preferable to hard-coded queries in the client code:

- Case 1: We change the schema in a way that makes these views impossible to reconstruct. In this case, we have a difficult migration regardless of whether the queries are stored (views) or hard-coded, since both would need to be manually fixed.
- Case 2: We change the schema in a way that changes _how_ the views are written, but does not change their surface area. In this case, the migration is fairly easy, as it only requires updating the schema and doesn't modify any data rows (or at least, not in a way we can see outside the dbms).
- Case 3: We change the schema in a way that reduces the surface area of the views. This case is actually the same whether we change views or real tables—any query coded against the that row structure would need to be updated.

If we have the client hard-code the view queries, then they have to update in all three of these cases. With the views, they only have to update in cases (1) and (3).

I don't foresee a need to do Case (1). Maybe if hypothetically we had to drop some of the fields for security reasons? But I'm having a hard time coming up with a realistic scenario for that, and extensions we can do without breaking them.

Case (2) seems more likely, but in that case any query written against the view should remain unchanged, perhaps modulo a data type change (which would have been breaking anyway).

Case (3) is possible but probably unnecessary except in circumstances similar to Case (1).

While it's true that changes to the schema _can_ sometimes be harder to manage than changes to client queries, in this case the data format we're using is intentionally pretty minimalistic. Clients certainly _can_ write their own more complex queries—and that's fine—all we're really doing is documenting which queries we think are important enough to defend against future changes (i.e., the views). 

In the end, the event sink is an internal log of the consensus engine, and I don't think we have to worry too much about future-proofing really esoteric abuses of it. If someone really wants to do something bizarre (beyond the features we've been asked to support here) they can always plug in their own indexer implementation.

That's how I see it, anyway—what do you think?

Reply:
I'm fine to keep the views, and I think my major concerns (upgrades, intelligibility, user experience) are well accounted for. I think I will, perhaps, always be concerned about "amount of database knowhow required to take advantage of database features," which I think is useful as a gut check but perhaps not always worth listening to.

Reply:
Thank you, I really appreciate your perspectives.